pkg/offsetguess/offsetguess.go

```diff
@@ -0,0 +1,304 @@
+	for i := 0; i < 16; i++ {
+		buf[i] = *(*byte)(unsafe.Pointer((uintptr(unsafe.Pointer(&ipv6Addr[0])) + uintptr(i))))
+	}
+	return net.IP(buf)
+}
+
+func htons(a uint16) uint16 {
+	arr := make([]byte, 2)
+	binary.BigEndian.PutUint16(arr, a)
+	return byteorder.Host.Uint16(arr)
+}
+
+// Guess expects elf.Module to hold a tcptracer-bpf object and initializes
+// the tracer by guessing the right kernel struct offsets. Results are
+// stored in the `tcptracer_status` map as used by the module.
+func Guess(b *elf.Module) error {
+	listenIP := "127.0.0.2"
+	listenPort := uint16(9091)
+	bindAddress := fmt.Sprintf("%s:%d", listenIP, listenPort)
+
+	finish := make(chan struct{})
+
+	currentNetns, err := ownNetNS()
+	if err != nil {
+		return fmt.Errorf("error getting current netns: %v", err)
+	}
+
+	mp := b.Map("tcptracer_status")
+
+	var zero uint64
+	pidTgid := uint64(os.Getpid()<<32 | syscall.Gettid())
+
+	status := tcpTracerStatus{
+		status:  checking,
+		pidTgid: pidTgid,
+	}
+
+	// if we already have the offsets, just return
+	err = b.LookupElement(mp, unsafe.Pointer(&zero), unsafe.Pointer(&status))
+	if err == nil && status.status == ready {
+		return nil
+	}
+
+	go listen(bindAddress, "tcp4", finish)
+	defer close(finish)
+	time.Sleep(300 * time.Millisecond)
+
+	if err := b.UpdateElement(mp, unsafe.Pointer(&zero), unsafe.Pointer(&status), 0); err != nil {
+		return fmt.Errorf("error initializing tcptracer_status map: %v", err)
+	}
+
+	dport := htons(listenPort)
+
+	// 127.0.0.1
+	saddr := 0x0100007F
+	// 127.0.0.2
+	daddr := 0x0200007F
+	// will be set later
+	sport := 0
+	netns := uint32(currentNetns)
+	family := syscall.AF_INET
+
+	for status.status != ready {
+		var daddrIPv6 [4]uint32
+
+		daddrIPv6[0] = rand.Uint32()
+		daddrIPv6[1] = rand.Uint32()
+		daddrIPv6[2] = rand.Uint32()
+		daddrIPv6[3] = rand.Uint32()
+
+		ip := ipFromUint32Arr(daddrIPv6)
+
+		if status.what != guessDaddrIPv6 {
+			conn, err := net.Dial("tcp4", bindAddress)
+			if err != nil {
+				return fmt.Errorf("error dialing %q: %v\n", bindAddress, err)
+			}
+
+			sport, err = strconv.Atoi(strings.Split(conn.LocalAddr().String(), ":")[1])
+			if err != nil {
+				return fmt.Errorf("error converting source port: %v", err)
+			}
+
+			sport = int(htons(uint16(sport)))
+
+			// set SO_LINGER to 0 so the connection state after closing is
+			// CLOSE instead of TIME_WAIT. In this way, they will disappear
+			// from the conntrack table after around 10 seconds instead of 2
+			// minutes
+			if tcpConn, ok := conn.(*net.TCPConn); ok {
+				tcpConn.SetLinger(0)
+			} else {
+				return fmt.Errorf("not a tcp connection unexpectedly")
+			}
+
+			conn.Close()
+		} else {
+			conn, err := net.Dial("tcp6", fmt.Sprintf("[%s]:9092", ip))
+			if err == nil {
+				conn.Close()
+			}
+		}
+
+		err = b.LookupElement(mp, unsafe.Pointer(&zero), unsafe.Pointer(&status))
+		if err != nil {
+			return fmt.Errorf("error reading tcptracer_status: %v", err)
+		}
+
+		if status.status == checked {
```

Comment:
You may want to choose documenting it on the C code, which is fine as long as you mention it here.